docs/contributing/contributing.rst

```diff
@@ -317,80 +317,69 @@ Maintainer's Guidelines
-  commits in those PRs.
-
-.. _here: http://chris.beams.io/posts/git-commit/
-.. _Waffle: https://waffle.io/BD2KGenomics/toil
+* **Commit messages** *should* be `great`_. Most importantly, they *must*:
+  
+  - Have a short subject line. If in need of more space, drop down **two** lines
+    and write a body to explain what is changing and why it has to change.
+  
+  - Write the subject line as a command: `Destroy all humans`,
+    not `All humans destroyed`.
+  
+  - Reference the issue being fixed in a Github-parseable format, such as
+    `(resolves #1234)` at the end of the subject line, or `This will fix #1234.`
+    somewhere in the body. If no single commit on its own fixes the issue, the
+    cross-reference must appear in the pull request title or body instead.
+    
+* **Branches** in the main Toil repository *must* start with ``issues/``,
+  followed by the issue number (or numbers, separated by a dash), followed by a
+  short, lowercase, hyphenated description of the change. (There can be many open
+  pull requests with their associated branches at any given point in time and
+  this convention ensures that we can easily identify branches.)
+  
+  Say there is an issue numbered #123 titled `Foo does not work`. The branch name
+  would be ``issues/123-fix-foo`` and the title of the commit would be
+  `Fix foo in case of bar (resolves #123).`
+  
+.. _great: https://chris.beams.io/posts/git-commit/#seven-rules
 
 Pull Requests
 ~~~~~~~~~~~~~
 * All pull requests must be reviewed by a person other than the request's
   author.
 
-* Only the reviewer of a pull request can merge it.
-
-* Until the pull request is merged, it should be continually rebased by the
-  author on top of master.
-
-* Pull requests are built automatically by Jenkins and won't be merged unless
-  all tests pass.
-
-* Ideally, a pull request should contain a single commit that addresses a
-  single, specific issue. Rebasing and squashing can be used to achieve that
-  goal (see :ref:`multi-author`).
-
-.. _multi-author:
-
-Multi-Author Pull Requests
-~~~~~~~~~~~~~~~~~~~~~~~~~~
-
-* A pull request starts off as single-author and can be changed to multi-author
-  upon request via comment (typically by the reviewer) in the PR. The author of
-  a single-author PR has to explicitly grant the request.
-
-* Multi-author pull requests can have more than one commit. They must `not` be
-  rebased as doing so would create havoc for other contributors.
-
-* To keep a multi-author pull request up to date with master, merge from master
-  instead of rebasing on top of master.
+* Pull requests will not be merged unless Travis and Gitlab CI tests pass.
+  Gitlab tests are only run on code in the main Toil repository on some branch,
+  so it is the responsibility of the approving reviewer to make sure that pull
+  requests from outside repositories are copied to branches in the main
+  repository. This can be accomplished with:
```

Comment:
To me this seems to be a sign that the UCSC gitlab is misconfigured. Look at the logs on Travis or Jenkins, they perform the `git fetch`ing needed automatically.

(also, this is super labor intensive and reduces our ability to accept outside contributions; a major anti-pattern! )

Reply:
I wouldn't say it's *misconfigured*; I think it's a design defect in Gitlab; it's a Git project hosting system first and a CI tool second, so it only knows how to do CI on projects that it itself is hosting, and we have to mirror Toil over to it to get things tested.

If we also mirrored over outside contributors' forks, that would work. And I'm sure we could write a bot to scan the Github PRs and automatically get the relevant commits into Gitlab, if we trust the author not to steal our environment variables (like the Jenkins whitelist). But as far as I can tell Gitlab has no notion of the existence of a PR on Github at all; all it knows how to do in its Github integration is to set the build status on a *commit*.

The relevant issue in Gitlab itself is https://gitlab.com/gitlab-org/gitlab-ee/issues/5667

If we figured this out before we set up Gitlab I probably would have recommended we just build another Jenkins instance instead. But now that we already have a vaguely working CI system, I'm not sure it would be worth it to switch back to simplify this workflow.